components/organisms/Chat/ChatMessages/CodeBlock.tsx

```diff
@@ -0,0 +1,93 @@
+import { Button } from "@/components/ui/Button";
+import { useContext, useEffect, useState } from "react";
+import { v4 as uuidv4 } from "uuid";
+import { ChatContext } from "./ChatContext";
+import { SalesforceQueryResultWithError } from "@/shared/types/salesforceTypes";
+
+type CodeBlockProps = {
+  code: string;
+  setQueryResult: (arg: SalesforceQueryResultWithError) => void;
+  queryResult: SalesforceQueryResultWithError;
+};
+
+export const CodeBlock = ({
+  code,
+  setQueryResult,
+  queryResult,
+}: CodeBlockProps) => {
+  const [isLoadingQuery, setIsLoadingQuery] = useState(false);
+  const { append, isLoading } = useContext(ChatContext);
+
+  const runQuery = async () => {
+    setIsLoadingQuery(true);
+    const salesforceId = localStorage.getItem("salesforceId");
+    const res = await fetch("/api/runQuery", {
+      method: "POST",
+      body: JSON.stringify({ salesforceId, query: code }),
+    });
+    const data = await res.json();
+    console.log(data);
+    const error = data?.error;
+    const result = data;
+    const instanceUrl = data?.instanceUrl;
+    const isRanBefore: boolean = Object.keys(queryResult ?? {}).length !== 0;
+    setQueryResult({ error, result, instanceUrl });
+    setIsLoadingQuery(false);
+    if (error && !isRanBefore) {
+      const newId = uuidv4();
+      await append({
+        id: `redo${newId}`,
+        createdAt: new Date(),
+        content:
+          "That last query didnt seem to work. PLease use a different strategy to give me a query that works. Note, if you are using " +
+          "custom fields/objects that haven't been mentioned in prior messages, please try something new. (use only standard objects/fields plus the custom objects/fields" +
+          " inputted in prior messages)",
+        role: "user",
+      });
+    }
+  };
+
+  useEffect(() => {
```

Comment:
Confused why this is needed

Reply:
This runs the query upon generation. So, after all the text is streamed in, i parse the text for the relevant code and run the query on it. However, sometimes i detect a code block before its already finished running, so i wait until its done loading, and then do it.